protocol/binary/envelope.go

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package binary implements the Thrift Binary protocol.
+package binary
+
+import (
+	"fmt"
+
+	"github.com/thriftrw/thriftrw-go/wire"
+)
+
+const (
+	versionMask = 0xffff0000
+	version1    = 0x80010000
+)
+
+// WriteEnveloped writes enveloped value using the strict envelope.
+// TODO: Add an option when creating the writer to choose non-strict writes.
+func (bw *Writer) WriteEnveloped(e wire.Envelope) error {
+	version := uint32(version1) | uint32(e.Type)
+
+	if err := bw.writeInt32(int32(version)); err != nil {
+		return err
+	}
+
+	if err := bw.writeString(e.Name); err != nil {
+		return err
+	}
+
+	if err := bw.writeInt32(e.SeqID); err != nil {
+		return err
+	}
+
+	return bw.WriteValue(e.Value)
+}
+
+// ReadEnveloped reads an Apache Thrift envelope
+//
+// Thrift supports two kinds of envelopes: strict, and non-strict.
+//
+// Non-strict envelopes:
+// Name (4 byte length prefixed string)
+// Type ID (1 byte)
+// Sequence ID (4 bytes)
+//
+// Strict envelopes:
+//
+// Version | Type ID (4 bytes)
+// Name (4 byte length prefixed string)
+// Sequence ID (4 bytes)
+//
+// When reading payloads, we need to support both strict and non-strict
+// payloads. To do this, we read the first 4 byte. Non-strict payloads
+// will always have a size >= 0, while strict payloads have selected
+// version numbers such that the value will always be negative.
+func (bw *Reader) ReadEnveloped() (wire.Envelope, error) {
+	var e wire.Envelope
+	initial, off, err := bw.readInt32(0)
```

Comment:
Haha, it is indeed overkill by 4x the amount of bytes we need. However, we pass the `initial` in to `readStrictNameType` which does need the rest of the bytes (rather than reading the 4 bytes again).

Mostly because at some point I'd like to move away from `io.ReaderAt` and make it just an `io.Reader`, where we can only consume bytes once.